### ISSUE TITLE
page renders before sagaTask is done

### ISSUE THREAD

**Issue by JerryCauser** [bmealhouse/next-redux-saga#3]
When we navigate through the pages "next-redux-saga" isn't waiting full loading of getInitialProps. This raises the problem of a "blinking page". When I click link (on gif it will be githab pages) on already loaded page - it show the page instantly, but data will load in 1-2 sec after that. It's not correct in most cases, bcs I wanna show the page with already loaded data. If we will load load page from server - everything is okay. I think we need await task end on client side too.
![loading_page](https://user-images.githubusercontent.com/5141037/34449439-2e18110a-ed09-11e7-9bab-8042577247b2.gif)

**Comment by bmealhouse** [comment]
Hi @JerryCauser, thanks for submitting this issue & PR.

I have the page loading immediately on purpose, to provide instant feedback to the user. Data that is loading asynchronously on the client should use a spinner or loading message to notify the client.

I'm not against providing an alternative way to handle routing with async data, however, I don't think this should replace the existing default behavior. Can you adjust your PR to support the existing default behavior, while allowing the ability to opt-in to this alternative rendering model? One way to do this would be using an named export for an alternative component.  Please make sure there are tests to support this as well.

Thoughts?

**Comment by JerryCauser** [comment]
> I'm not against providing an alternative way to handle routing with async data, however, I don't think this should replace the existing default behavior.

Yeah, of course. Backwards compatibility in priority. I accent on that trouble because redux-observable, redux-thunk and redux-promise-middleware (and all "promise like" middlewares) have same scenario of lifecycle with awaiting on client side as I decrived before.


> Can you adjust your PR to support the existing default behavior, while allowing the ability to opt-in to this alternative rendering model?
Yeah, of course. I have also 2 ideas in addition to yours. 
1. Make @decorator `withReduxSaga` smarter and give him possibility to get async/sync flag as first argument. Something like that:
```javascript
//async client side
class MyComponent extends React.Component {/**some code**/}
export const wrappedComponent = withReduxSaga(MyComponent)

//sync client side
class MyComponent extends React.Component {/**some code**/}
export const wrappedComponent = withReduxSaga({sync: true})(MyComponent)

//with es6-decorators
//async client side
@ withReduxSaga
class MyComponent extends React.Component {/**some code**/}

//sync client side
@ withReduxSaga({sync: true})
class MyComponent extends React.Component {/**some code**/}
```

2. Let's leave the decorator alone. And let component describe itself by static props `sync`:
```javascript
//async client side
@ withReduxSaga
class MyComponent extends React.Component {/**some code**/} 

//sync client side
@ withReduxSaga
class MyComponent extends React.Component {
  static sync = true //sync by default undefined, whats mean false for backwards compatibility
  /**some code**/
}
```

Which one look better for you? I like idea with tuning @decorator, but idea with second decorator (`withReduxSagaSync` for example) is nice too.

> Please make sure there are tests to support this as well.

I think here I will need your advice which test I should add. But anyway it's not the problem.

**Comment by bmealhouse** [comment]
> I accent on that trouble because redux-observable, redux-thunk and redux-promise-middleware (and all "promise like" middlewares) have same scenario of lifecycle with awaiting on client side as I described before.

Thanks again for pointing this out.  I did some more research on how Next.js works out of the box and in their data-fetch example, it does wait for `getInitialProps` to finish loading before transitioning to the new route.  If that is how the other projects you mentioned work as well, I think it's better to break the API for `next-redux-saga` to be consistent with the community.

That being said, what do you think about releasing a new major version (2.x) with sync as the default?  Passing `{async: true}` would provide backwards compatibility.

```javascript

// sync client side
class MyComponent extends React.Component {/**some code**/}
export const wrappedComponent = withReduxSaga(MyComponent)

// async client side
class MyComponent extends React.Component {/**some code**/}
export const wrappedComponent = withReduxSaga({async: true})(MyComponent)

// with es6-decorators
// sync client side
@ withReduxSaga
class MyComponent extends React.Component {/**some code**/}

// async client side
@ withReduxSaga({async: true})
class MyComponent extends React.Component {/**some code**/}
```

> I think here I will need your advice which test I should add. But anyway it's not the problem.

No problem, we can work through the tests together.

**Comment by JerryCauser** [comment]
done, sorry for the delay. 
check the tests (are they ok?) and review pr https://github.com/bmealhouse/next-redux-saga/pull/4

**Comment by bmealhouse** [comment]
Thanks for your work @JerryCauser, this is great!  Closed by #4.

There is also bunch of cleanup I need to do on this repo before publishing the 2.0.0 release.  I will try to get those things done in the next couple days.

**Comment by bmealhouse** [comment]
@JerryCauser v2.x is now published! 🎉 

Thanks again for your contribution.
